lib/wal/wal.c

```diff
@@ -90,6 +91,66 @@ struct wal_sync_waiter {
 void
 wal_ionotify_cb(void *cbarg, merr_t err);
 
+HSE_ALWAYS_INLINE static void
+wal_throttle_sensor_set(struct wal *wal, uint64_t bufsz, uint64_t buflen)
+{
+    const uint64_t hwm = (bufsz * wal->wal_thr_hwm) / 100;
+    const uint64_t lwm = (bufsz * wal->wal_thr_lwm) / 100;
+    uint32_t new;
+
+    if (ev(!wal->wal_thr_sensor))
+        return;
+
+    ev(buflen >= bufsz);
+
+    new = (buflen > lwm) ? (THROTTLE_SENSOR_SCALE * buflen) / hwm : 0;
+
+    throttle_sensor_set(wal->wal_thr_sensor, new);
+}
+
+/*
+ * Wait for at least 'pct' of flushed data to become durable.
+ * This avoids overloading the WAL IO layer with numerous flushes when IOs are
+ * backed up due to a slow IO backend.
+ */
+HSE_ALWAYS_INLINE static void
+wal_flush_wait(struct wal *wal, struct wal_flush_stats *stats, uint8_t pct)
+{
+    INVARIANT(wal && stats);
+
+    const int flushc = stats->bufcnt;
+    uint64_t flushv[flushc];
+
+    assert(flushc > 0);
+    for (int i = 0; i < flushc; i++)
+        flushv[i] = stats->flush_soff[i] + ((stats->flush_len[i] * pct) / 100);
+
+    while (wal_bufset_durcnt(wal->wbs, flushc, flushv) < flushc)
+        usleep(50);
+}
+
+HSE_ALWAYS_INLINE static bool
+wal_dirty_exceeds_threshold(struct wal *wal, const int flushc, uint64_t *flushv)
+{
+    INVARIANT(wal && flushc > 0);
+
+    uint64_t curv[flushc], tot_bytes = 0;
+    uint32_t buf_thresh = wal->dur_bytes / flushc;
+    uint8_t  buf_cnt = 0;
```

Comment:
I think it'd be better not to use VLA for these flushc arrays (and in general we should avoid VLAs).  We have a max, right?
Also, please don't use 8-bit ints when a uint would do, there's no upside (no perf or space benefit, and can actually hinder perf), not to mention I keep running into issues where some code uses an 8-bit int which silently rolls over when I modify the code to do something new, PITA to debug.  Generally I would use a simple int unless there's a critical need to use something smaller (like in a structure to avoid bloat).

Reply:
Agreed and fixed...

Reply:
There is a flag to warn about VLAs that we should probably enable at some point, but I don't really want to deal with all the warnings.